java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -847,7 +919,7 @@ private module ControlFlowGraphImpl {
       // any other abnormal completion is propagated
       last(switch.getAStmt(), last, completion) and
       completion != anonymousBreakCompletion() and
-      completion != NormalCompletion()
+      not completion instanceof NormalOrBooleanCompletion
       or
       // if the last case completes normally, then so does the switch
       last(switch.getStmt(strictcount(switch.getAStmt()) - 1), last, NormalCompletion()) and
```

Comment:
In the lines below here (which I can't comment on since they're too far from the diff), which read
```
      // if no default case exists, then normal completion of the expression may terminate the switch
      not exists(switch.getDefaultCase()) and
      last(switch.getExpr(), last, completion) and
      completion = NormalCompletion()
```
we'll need to add `not exists(switch.getAPatternCase()), since we're doing sequential testing of the pattern cases and therefore shouldn't include the "nothing matched" edge.

Reply:
In fact the JLS requires exhaustiveness if either null or pattern cases are involved (i.e., we're sure this isn't a grandfathered pre-Java-21 switch block), so we know such an edge is impossible if the program compiles.